### PR TITLE
Feature: NewGRF callback profiling

### PR DESCRIPTION
**Pull request by nielsmh**
Adds a console command `newgrf_profile` to collect some profiling data about NewGRF action 2 callbacks and produce a CSV file.

I'm not entirely sure if this is useful or not, hopefully someone can try it out and decide. It should have almost no impact on performance when not in use. The console command requires developer tools to be enabled.

Careful, the CSV files can get very large very fast :) One small-ish test game with (a beta of) Iron Horse being profiled made a CSV file of 8 MB with more than 460k events recorded across 7 in-game days.

### CODE REVIEW

**Review by planetmaker — DISMISSED** [review]
Such feature will totally be helpful for more complex NewGRFs and their profiling :)

**Comment by andythenorth** [comment]
Just extra info...

CB IDs: https://github.com/OpenTTD/OpenTTD/blob/master/src/newgrf_callbacks.h#L20

Filesize: 30 days of Horse profiling (in a small game) gives 3m line csv.  Some Excel versions, Apple Numbers etc won't open large files (the limit is sometimes 65k).  Libre Office, some Excel versions might do more.  I'm thinking that grf authors will find it better to write a python script to summarise the data, or maybe a jupyter notebook or similar.

**Comment by planetmaker** [comment]
Yes, but the deficiencies of spreadsheet programmes should not limit us in our debug output :)

File: src/console_cmds.cpp [hunk 1]
```diff
@@ -1872,6 +1873,51 @@ DEF_CONSOLE_CMD(ConNewGRFReload)
 	return true;
 }
 
+DEF_CONSOLE_CMD(ConNewGRFProfile)
+{
+	if (argc == 0) {
+		IConsoleHelp("Collect a callback profiling session from a NewGRF for a number of in-game days.");
+		IConsoleHelp("Usage: newgrf_profile [<grf-num> <num-days>]");
+		IConsoleHelp("Run without parameters to get a list of active NewGRFs and their grf-num");
+		return true;
```
Comment (frosch123) [inline src/console_cmds.cpp]: If debugging multiple grfs would be a thing, one could imagine these commands:
* Show status: newgrf_profile
* Add GRF to profiler: newgrf_profile enable < grf-num >
* Remove GRF from profiler: newgrf_profile disable < grf-num >
* Run profiler for N days: newgrf_profile start < days >
* Run profiler indefinitely: newgrf_profile start
* Stop profiler: newgrf_profile stop

File: src/date.cpp [hunk 2]
```diff
@@ -243,6 +244,11 @@ static void OnNewMonth()
  */
 static void OnNewDay()
 {
+	if (_newgrf_profiler != nullptr && _newgrf_profiler->profile_end_date <= _date) {
+		delete _newgrf_profiler;
```
Comment (frosch123) [inline src/date.cpp]: this suggests that _newgrf_profiler should be a unique_ptr

File: src/newgrf.cpp [hunk 3]
```diff
@@ -66,6 +66,11 @@
 /** List of all loaded GRF files */
 static std::vector<GRFFile *> _grf_files;
 
+std::vector<GRFFile *> GetAllGRFFiles()
+{
```
Comment (frosch123) [inline src/newgrf.cpp:70]: return "const std::vector<>&"?

File: src/newgrf_profiling.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+ */
+NewGRFProfiler::~NewGRFProfiler()
+{
+	std::string filename = this->GetOutputFilename();
+	IConsolePrintF(TC_LIGHT_BROWN, "Finished profile of NewGRF, writing %u events to %s", (uint)this->calls.size(), filename.c_str());
+
+	FILE *f = FioFOpenFile(filename.c_str(), "wt", Subdirectory::NO_DIRECTORY);
+	FileCloser fcloser(f);
+
+	fputs("Tick,Sprite,CallbackID,Microseconds,Subs\n", f);
```
Comment (frosch123) [inline src/newgrf_profiling.cpp]: "GRFID", "feature" and "GRF-local" item id would be useful

File: src/newgrf_profiling.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	void BeginResolve(const ResolverObject &resolver);
+	void EndResolve();
+	void RecursiveResolve();
+
+	std::string GetOutputFilename() const;
+
+	/** Measurement of a single sprite group resolution */
+	struct Call {
+		uint32 nfo_line; ///< Pseudo-sprite index in GRF file
+		uint32 subs;     ///< Sub-calls to other sprite groups
```
Comment (frosch123) [inline src/newgrf_profiling.h]: "subs" sounds as if it is about procedure calls.
Maybe "num_spritegroups"?

**Review by frosch123 — DISMISSED** [review]

File: src/newgrf_profiling.cpp [hunk 6]
```diff
@@ -64,10 +64,21 @@ void NewGRFProfiler::BeginResolve(const ResolverObject &resolver)
...
 {
 	using namespace std::chrono;
 	this->cur_call.time = (uint32)time_point_cast<microseconds>(high_resolution_clock::now()).time_since_epoch().count() - this->cur_call.time;
+
+	if (result == nullptr) {
+		this->cur_call.result = 0;
+	} else if (result->type == SGT_CALLBACK) {
+		this->cur_call.result = static_cast<const CallbackResultSpriteGroup *>(result)->result;
+	} else if (result->type == SGT_RESULT) {
+		this->cur_call.result = static_cast<const ResultSpriteGroup *>(result)->sprite;
```
Comment (frosch123) [inline src/newgrf_profiling.cpp]: This is a global sprite id.
"GetSpriteCache(sprite).id" has the local id.

File: docs/logging_and_performance_metrics.md [hunk 7]
```diff
@@ -99,3 +99,40 @@ The following is an explanation of the different statistics:
...
+
+NewGRF developers can profile callback chains via the `newgrf_profile`
+console command. The command enables a profiling mode where every sprite
+request is measured and logged, and written to a CSV file in the end.
+
+Usage:
+
+- `newgrf_profile` - show a list of loaded GRFs and their index numbers.
+- `newgrf_profile <grf-num> <days>` - Begin profiling the GRF and run profiling
+  for that many in-game days. Automarically writes a CSV file to the screenshot
```
Comment (glx22) [inline docs/logging_and_performance_metrics.md]:
```suggestion
  for that many in-game days. Automatically writes a CSV file to the screenshot
```

File: src/newgrf_profiling.cpp (hunk 4 above)
Comment (nielsmh) [reply src/newgrf_profiling.cpp]: Started adding this, just not for all the various resolvers yet.

File: src/newgrf_canal.cpp [hunk 8]
```diff
@@ -111,6 +116,16 @@ struct CanalResolverObject : public ResolverObject {
...
 }
 
+GrfSpecFeature CanalResolverObject::GetFeature() const
+{
+	return GSF_CANALS;
+}
+
+uint32 CanalResolverObject::GetLocalID() const
+{
+	return 0; // TODO? not sure there is one/this is possible
```
Comment (frosch123) [inline src/newgrf_canal.cpp]: yes, it's the "CanalFeature feature" member you already added.

File: src/newgrf_cargo.cpp [hunk 9]
```diff
@@ -30,6 +35,16 @@ struct CargoResolverObject : public ResolverObject {
...
 }
 
+GrfSpecFeature CargoResolverObject::GetFeature() const
+{
+	return GSF_CARGOES;
+}
+
+uint32 CargoResolverObject::GetLocalID() const
+{
+	return this->cargospec->label; // Not quite the same as ID, but should work
```
Comment (frosch123) [inline src/newgrf_cargo.cpp]: Technically the local id is "cargospec->Index()". But I agree that cargoslots and cargobits are confusing and "label" is more useful.
However, the method should not be named "GetLocalID" then.

File: src/newgrf_railtype.cpp [hunk 10]
```diff
@@ -65,6 +65,16 @@
...
 }
 
+GrfSpecFeature RailTypeResolverObject::GetFeature() const
+{
+	return GSF_RAILTYPES;
+}
+
+uint32 RailTypeResolverObject::GetLocalID() const
+{
+	return this->railtype_scope.rti->label;
```
Comment (frosch123) [inline src/newgrf_railtype.cpp:75]: The local id is super weird to get (reverse lookup in railtype_map), and is actually not used anywhere after initialisation. So I agree that "label" is more useful.
However, the method should not be named "GetLocalID" then.

File: src/newgrf_roadtype.cpp [hunk 11]
```diff
@@ -65,16 +65,27 @@
...
 }
 
+GrfSpecFeature RoadTypeResolverObject::GetFeature() const
+{
+	return GSF_ROADTYPES; // FIXME: when is this GSF_TRAMTYPES instead?
+}
+
+uint32 RoadTypeResolverObject::GetLocalID() const
+{
+	return this->roadtype_scope.rti->label;
```
Comment (frosch123) [inline src/newgrf_roadtype.cpp:80]: The local id is super weird to get (reverse lookup in road/tramtype_map), and is actually not used anywhere after initialisation. So I agree that "label" is more useful.
However, the method should not be named "GetLocalID" then.

**Comment by frosch123** [comment]
How about "GetDebugID"?

File: src/newgrf.cpp (hunk 3 above)
Comment (nielsmh) [reply src/newgrf.cpp:70]: I'm considering just make the variable non-static instead.

Comment (frosch123) [reply src/newgrf.cpp:70]: Another option is to pass the grfid instead of some enumeration of the loaded grfs.
Then it's the same every time.

File: src/newgrf_spritegroup.cpp [hunk 12]
```diff
@@ -36,17 +36,20 @@ TemporaryStorageArray<int32, 0x110> _temp_store;
 {
 	if (group == nullptr) return nullptr;
 
-	if (_newgrf_profiler == nullptr || _newgrf_profiler->grffile != object.grffile) {
+	const GRFFile *grf = object.grffile;
+	auto profiler = std::find_if(_newgrf_profilers.begin(), _newgrf_profilers.end(), [&](const NewGRFProfiler &pr) { return pr.grffile == grf; });
```
Comment (frosch123) [inline src/newgrf_spritegroup.cpp:40]: This can only change if top_level==true, so it may be worth caching somewhere.

File: src/console_cmds.cpp [hunk 13]
```diff
@@ -1877,44 +1877,120 @@ DEF_CONSOLE_CMD(ConNewGRFProfile)
 {
 	if (argc == 0) {
 		IConsoleHelp("Collect a callback profiling session from a NewGRF for a number of in-game days.");
-		IConsoleHelp("Usage: newgrf_profile [<grf-num> <num-days>]");
-		IConsoleHelp("Run without parameters to get a list of active NewGRFs and their grf-num");
+		IConsoleHelp("Usage: newgrf_profile [list]");
+		IConsoleHelp("  List all NewGRFs that can be profiled, and their status.");
+		IConsoleHelp("Usage: newgrf_profile add <grf-num>...");
```
Comment (frosch123) [inline src/console_cmds.cpp]: The "content" command uses "select" and "unselect". Does it make sense to use the same terms?

File: src/console_cmds.cpp [hunk 14]
```diff
@@ -1877,44 +1877,120 @@ DEF_CONSOLE_CMD(ConNewGRFProfile)
...
+		IConsoleHelp("  Select one or more GRFs for profiling.");
+		IConsoleHelp("Usage: newgrf_profile rem <grf-num>...");
+		IConsoleHelp("  Deselect one or more GRFs from profiling. Use the keyword \"all\" instead of a GRF number to remove all.");
```
Comment (frosch123) [inline src/console_cmds.cpp]: Also mention that this "aborts" running profilers?

File: docs/logging_and_performance_metrics.md [hunk 15]
```diff
@@ -102,22 +102,32 @@ current simulation rate and the game speed factor.
 ## 3.0) NewGRF callback profiling
 
 NewGRF developers can profile callback chains via the `newgrf_profile`
-console command. The command enables a profiling mode where every sprite
+console command. The command controls a profiling mode where every sprite
 request is measured and logged, and written to a CSV file in the end.
 
+The NewGRF developer tools need to be enabled for the command to function.
+
 Usage:
```
Comment (frosch123) [inline docs/logging_and_performance_metrics.md]: Maybe just reference the ingame help here, instead of duplicating the text.

File: src/console_cmds.cpp (hunk 13 above)
Comment (nielsmh) [reply src/console_cmds.cpp]: Eh I guess that's doable. I'd rather not have long sub-command names when the base command is already long and has an underscore, but can allow abbreviations (just prefix match on "sel" and "uns"?)

File: src/newgrf_spritegroup.cpp (hunk 12 above)
Comment (nielsmh) [reply src/newgrf_spritegroup.cpp:40]: I don't think it's worth optimising this, for regular gameplay the profilers list will be empty so in -O2 builds the entire find_if hopefully compiles to have a fast is-empty path, and if the player is operating the profiler they will have to put up with a small bit of extra slowness.

File: src/console_cmds.cpp [hunk 16]
```diff
@@ -1872,6 +1873,135 @@ DEF_CONSOLE_CMD(ConNewGRFReload)
...
+		}
+		return true;
+	}
+
+	/* "select" sub-command */
+	if (strncasecmp(argv[1], "sel", 3) == 0 && argc >= 3) {
+		for (size_t argnum = 2; argnum < argc; ++argnum) {
+			int grfnum = atoi(argv[argnum]);
+			if (grfnum < 1 || grfnum > (int)files.size()) { // safe cast, files.size() should not be larger than a few hundred in the most extreme cases
+				IConsolePrintF(TC_YELLOW, "GRF number %d out of range, not added.", grfnum);
```
Comment (frosch123) [inline src/console_cmds.cpp]: CC_INFO or CC_WARNING?

File: src/console_cmds.cpp [hunk 17]
```diff
@@ -1872,6 +1873,135 @@ DEF_CONSOLE_CMD(ConNewGRFReload)
...
+				continue;
+			}
+			GRFFile *grf = files[grfnum - 1];
+			if (std::any_of(_newgrf_profilers.begin(), _newgrf_profilers.end(), [&](NewGRFProfiler &pr) { return pr.grffile == grf; })) {
+				IConsolePrintF(TC_YELLOW, "GRF number %d [%08X] is already selected for profiling.", grfnum, BSWAP32(grf->grfid));
```
Comment (frosch123) [inline src/console_cmds.cpp]: CC_INFO or CC_WARNING?

File: src/console_cmds.cpp [hunk 18]
```diff
@@ -1872,6 +1873,135 @@ DEF_CONSOLE_CMD(ConNewGRFReload)
...
+	/* "unselect" sub-command */
+	if (strncasecmp(argv[1], "uns", 3) == 0 && argc >= 3) {
+		for (size_t argnum = 2; argnum < argc; ++argnum) {
+			if (strcasecmp(argv[argnum], "all") == 0) {
+				_newgrf_profilers.clear();
+				break;
+			}
+			int grfnum = atoi(argv[argnum]);
+			if (grfnum < 1 || grfnum > (int)files.size()) {
+				IConsolePrintF(TC_YELLOW, "GRF number %d out of range, not removing.", grfnum);
```
Comment (frosch123) [inline src/console_cmds.cpp]: CC_INFO or CC_WARNING?

File: src/console_cmds.cpp [hunk 19]
```diff
@@ -1872,6 +1873,135 @@ DEF_CONSOLE_CMD(ConNewGRFReload)
...
+				started++;
+
+				if (!grfids.empty()) grfids += ", ";
+				char grfidstr[12]{ 0 };
+				seprintf(grfidstr, lastof(grfidstr), "[%08X]", BSWAP32(pr.grffile->grfid));
+				grfids += grfidstr;
+			}
+		}
+		if (started > 0) {
+			IConsolePrintF(TC_LIGHT_BROWN, "Started profiling for GRFID%s %s", (started > 1) ? "s" : "", grfids.c_str());
```
Comment (frosch123) [inline src/console_cmds.cpp]: CC_DEBUG

File: src/console_cmds.cpp [hunk 20]
```diff
@@ -1872,6 +1873,135 @@ DEF_CONSOLE_CMD(ConNewGRFReload)
...
+			if (argc >= 3) {
+				int days = max(atoi(argv[2]), 1);
+				_newgrf_profile_end_date = _date + days;
+
+				char datestrbuf[32]{ 0 };
+				SetDParam(0, _newgrf_profile_end_date);
+				GetString(datestrbuf, STR_JUST_DATE_ISO, lastof(datestrbuf));
+				IConsolePrintF(TC_LIGHT_BROWN, "Profiling will automatically stop on game date %s", datestrbuf);
```
Comment (frosch123) [inline src/console_cmds.cpp]: CC_DEBUG

File: src/console_cmds.cpp [hunk 21]
```diff
@@ -1872,6 +1873,135 @@ DEF_CONSOLE_CMD(ConNewGRFReload)
...
+			} else {
+				_newgrf_profile_end_date = MAX_DAY;
+			}
+		} else if (_newgrf_profilers.empty()) {
+			IConsolePrintF(TC_YELLOW, "No GRFs selected for profiling, did not start.");
```
Comment (frosch123) [inline src/console_cmds.cpp]: CC_INFO or CC_WARNING?

File: src/console_cmds.cpp [hunk 22]
```diff
@@ -1872,6 +1873,135 @@ DEF_CONSOLE_CMD(ConNewGRFReload)
...
+		} else {
+			IConsolePrintF(TC_YELLOW, "Did not start profiling for any GRFs, all selected GRFs are already profiling.");
```
Comment (frosch123) [inline src/console_cmds.cpp]: CC_INFO or CC_WARNING?

File: src/console_cmds.cpp [hunk 23]
```diff
@@ -1872,6 +1873,135 @@ DEF_CONSOLE_CMD(ConNewGRFReload)
...
+
+	/* "list" sub-command */
+	if (argc == 1 || strncasecmp(argv[1], "lis", 3) == 0) {
+		IConsolePrint(TC_LIGHT_BROWN, "Loaded GRF files:");
+		int i = 1;
+		for (GRFFile *grf : files) {
+			auto profiler = std::find_if(_newgrf_profilers.begin(), _newgrf_profilers.end(), [&](NewGRFProfiler &pr) { return pr.grffile == grf; });
+			bool selected = profiler != _newgrf_profilers.end();
+			bool active = selected && profiler->active;
+			TextColour tc = active ? TC_GREEN : selected ? TC_LIGHT_BLUE : TC_LIGHT_BROWN;
```
Comment (frosch123) [inline src/console_cmds.cpp]: The "content" command uses different colors for selected items. Though they are argueably weird and use CC_xxx constants out of their scope.
I think there should be separate CC_xxx constants for selected/unselected/... items, which should be shared between profiler and content.

File: src/console_cmds.cpp [hunk 24]
```diff
@@ -1872,6 +1873,135 @@ DEF_CONSOLE_CMD(ConNewGRFReload)
...
+		IConsoleHelp("  End profiling and discard all collected data.");
+		return true;
+	}
+
+	extern const std::vector<GRFFile *> &GetAllGRFFiles();
+	const std::vector<GRFFile *> &files = GetAllGRFFiles();
+
+	/* "list" sub-command */
+	if (argc == 1 || strncasecmp(argv[1], "lis", 3) == 0) {
+		IConsolePrint(TC_LIGHT_BROWN, "Loaded GRF files:");
```
Comment (frosch123) [inline src/console_cmds.cpp]: "content" command used CC_WHITE for headers.

File: src/console_cmds.cpp (hunk 23 above)
Comment (nielsmh) [reply src/console_cmds.cpp]: I'd rather not touch the content command in this PR, but maybe later.

**Review by LordAro — APPROVED** [review]
Code looks fine, would like to get @frosch123's OK first though...

**Comment by frosch123** [comment]
I experimented a bit with this last week.
I found the output quite hard to parse, due to the weird wrap-around behaviour of ticks.
For vehicles the "item id" was some random noise that I did not recognise.

**Comment by nielsmh** [comment]
I think most NML GRFs will have automatically assigned item id's, but it really ought to be the value you can use to figure out which specific engine, industry tile, or whatever, that's causing the activity.
Perhaps the ticks should be taken as difference from the data collection start, so the first measurement is always at zero? It might even be possible to handle integer wraparound like that, in case someone actually captures more than 64k ticks worth of events.

**Comment by nielsmh** [comment]
I'm unsure if it's okay to squash this to one commit (it's a large changeset), or if I should put in the effort to try make logical commits. I don't think there is much to separate out.
